service/testlog/handlers_testlog.go

```diff
@@ -0,0 +1,123 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ * License granted by Canonical Limited
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package testlog
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"mime/multipart"
+	"net/http"
+	"os"
+	"text/template"
+	"time"
+
+	"github.com/CanonicalLtd/serial-vault/service/response"
+)
+
+const tpl = `
+<!DOCTYPE html>
+<html>
+	<head>
+		<title>Test Log Upload</title>
+		<style>
+		body {font-family: Ubuntu,Roboto,sans-serif}
+		h1   {margin-top: 0; padding: 10px 10px; background-color: #E95420; color: #fff; font-weight: normal}
+		div { margin: 40px}
+		fieldset { padding: 20px}
+		</style>
+    </head>
+    <body>
+		<h1>Serial Vault</h1>
+		<div>
+			<h2>Test Log Upload</h2>
+			<form action="/testlog" method="POST" enctype="multipart/form-data">
+				<fieldset>
+				<label for="logfile">Filename:</label><br />
+				<input type="file" name="logfile" accept="text/xml">
+				<input type="Submit">
+				</fieldset>
+			</form>
+		</div>
+	</body>
+</html>
+`
+const paramsEnvVar = "SNAP_DATA"
+
+// Index is the form of the test log upload web application
+func Index(w http.ResponseWriter, r *http.Request) {
+	t, err := template.New("testlog").Parse(tpl)
+	if err != nil {
+		log.Printf("Error loading the application template: %v\n", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	err = t.Execute(w, nil)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+}
+
+// Submit is the POST request for caching a test log
+func Submit(w http.ResponseWriter, r *http.Request) {
+	file, handle, err := r.FormFile("logfile")
+	if err != nil {
+		formatUploadResponse(w, http.StatusBadRequest, err.Error())
+		return
+	}
+	defer file.Close()
+
+	saveFile(w, file, handle)
+}
+
+func formatUploadResponse(w http.ResponseWriter, code int, message string) {
+	w.Header().Set("Content-Type", response.JSONHeader)
+	w.WriteHeader(code)
+	fmt.Fprint(w, message)
+}
+
+func saveFile(w http.ResponseWriter, file multipart.File, handle *multipart.FileHeader) {
+	data, err := ioutil.ReadAll(file)
+	if err != nil {
+		formatUploadResponse(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	// Set up the path for the file
+	path := os.Getenv(paramsEnvVar)
+	if len(path) == 0 {
+		path = "."
+	}
+	path = path + "/files"
+	_ = os.Mkdir(path, 0755)
```

Comment:
unlikely but still shouldn't it be logged when mkdir fails - like config variable points to a not writable location

Reply:
It will fail if the directory already exists as well. This will only be used within a snap, so I think its fine for now. I could do a check if directory-exists-and-create-it-if-not, but I think this is a little for efficient.